tests/test_package.py

```diff
@@ -68,6 +68,16 @@ def fn_getter():
             with self.assertRaises(IOError):
                 pkg._header
 
+    @mock.patch("rpm.TransactionSet.dbMatch", lambda self, a, b: iter([rpm.hdr()]))
+    def test_get_header(self):
+        pkg = self.sack.query().installed().filter(name="pepper")[0]
+        header = pkg.get_header()
+        self.assertIsInstance(header, rpm.hdr)
```

Comment:
I suggest that the test does not test what is supposed to. It tests `a == a`, but it is hidden. Or in case that original function `dbMatch(0, self.rpmdbid)` will crash or return something else, the test will still pass, because call `rpm.TransactionSet.dbMatch` is replaces by `lambda self, a, b: iter([rpm.hdr()])`. And as I know `a == a` always pass. Is there any chance to test it by a different way?

Reply:
@j-mracek this is the best I could come up with. You're right that it's not perfect, but I think it does test the actual body of the `get_header()` method, in the sense that it checks it returns an instance of the correct class when the package is installed and `None` when the package isn't... I don't see a way to do this better.

Reply:
The solution would be to use real data in test and not Mocked packages. See `test_modules.py`

Reply:
Anyway it looks like that a proper setup from test_modules.py will not help because we need in installroot installed content. It means it will require to install rpm in installroot first.

Reply:
I provide a test that do not use Mocking. This is only example that should be improved (replace subprocess by rpm api, and there should be a test that returned header has correct content, like name ...).
```
from __future__ import absolute_import
from __future__ import unicode_literals

import os
import shutil
import subprocess
import tempfile
import unittest

import rpm

import dnf.conf
import dnf.base

TOP_DIR = os.path.abspath(os.path.dirname(__file__))
REPOS_DIR = os.path.join(TOP_DIR, "modules/modules")


class PackageTest(unittest.TestCase):
    def setUp(self):
        self.tmpdir = tempfile.mkdtemp(prefix="dnf_test_")
        rpm1_path = os.path.join(
            TOP_DIR,"modules/modules/_non-modular/x86_64/httpd-provides-name-doc-3.0-1.x86_64.rpm")
        rpm2_path = os.path.join(
            TOP_DIR, "modules/modules/_non-modular/x86_64/httpd-doc-2.2.10-1.x86_64.rpm")
        command = ['rpm', '--root', self.tmpdir, '-i', '--nosignature', rpm1_path, rpm2_path]
        subprocess.check_call(command)
        self.conf = dnf.conf.MainConf()
        self.conf.cachedir = os.path.join(self.tmpdir, "cache")
        self.conf.installroot = os.path.join(self.tmpdir)
        self.conf.persistdir = os.path.join(self.conf.installroot, self.conf.persistdir.lstrip("/"))
        self.conf.substitutions["arch"] = "x86_64"
        self.conf.substitutions["basearch"] = dnf.rpm.basearch(self.conf.substitutions["arch"])
        self.base = dnf.Base(conf=self.conf)

        self._add_module_repo("_non-modular")
        self.base.fill_sack(load_system_repo=True)

    def tearDown(self):
        shutil.rmtree(self.tmpdir)

    def _add_module_repo(self, repo_id, modules=True):
        url = "file://" + os.path.join(REPOS_DIR, repo_id, self.conf.substitutions["arch"])
        repo = self.base.repos.add_new_repo(repo_id, self.base.conf, baseurl=[url], modules=modules)
        return repo

    def test_get_header(self):
        pkg = self.base.sack.query().installed().filter(name="httpd-provides-name-doc")[0]
        header = pkg.get_header()
        self.assertIsInstance(header, rpm.hdr)

        pkg = self.base.sack.query().available().filter(name="grub2")[0]
        header = pkg.get_header()
        self.assertEqual(header, None)
```

Reply:
@j-mracek I've replaced the empty header in the mocked method with a "unique" string. I don't see an easy way to mock an `rpm.hdr` instance with custom data... PTAL.

Reply:
Thanks, it is a better now.